x/staking/keeper/delegation.go

```diff
@@ -791,19 +822,22 @@ func (k Keeper) CompleteUnbonding(ctx sdk.Context, delAddr sdk.AccAddress, valAd
 	for i := 0; i < len(ubd.Entries); i++ {
 		entry := ubd.Entries[i]
 		if entry.IsMature(ctxTime) {
-			ubd.RemoveEntry(int64(i))
-			i--
-
-			// track undelegation only when remaining or truncated shares are non-zero
-			if !entry.Balance.IsZero() {
-				amt := sdk.NewCoin(bondDenom, entry.Balance)
-				if err := k.bankKeeper.UndelegateCoinsFromModuleToAccount(
-					ctx, types.NotBondedPoolName, delegatorAddress, sdk.NewCoins(amt),
-				); err != nil {
-					return nil, err
+			// call hook, if error is returned, stop unbonding
+			err := k.BeforeUnbondingDelegationEntryComplete(ctx, entry.Id)
+			if err == nil {
```

Comment:
This will keep the entries in the queue, and just reiterate over the same entries correct?